devito/types/dense.py

```diff
@@ -1360,6 +1361,31 @@ def __shape_setup__(cls, **kwargs):
                 shape.insert(cls._time_position, save)
             else:
                 raise TypeError("`save` can be None, int or Buffer, not %s" % type(save))
+        elif dimensions is None:
+            raise TypeError("`dimensions` required if both `grid` and "
+                            "`shape` are provided")
+        else:
+            # Got `grid`, `dimensions`, and `shape`. We sanity-check that the
+            # Dimensions in `dimensions` also appearing in `grid` have same size
+            # (given by `shape`) as that provided in `grid`
+            if len(shape) != len(dimensions):
+                raise ValueError("`shape` and `dimensions` must have the "
+                                 "same number of entries")
+            loc_shape = []
+            for d, s in zip(dimensions, shape):
+                if d in grid.dimensions:
+                    size = grid.dimension_map[d]
+                    if size.glb != s and s is not None:
+                        raise ValueError("Dimension `%s` is given size `%d`, "
```

Comment:
`InputError` may be more adapted.

Reply:
Again, have lifted from `Function` but happy to change both (as a second commit). Let me know.